File: kernel/ipc/mutex.c
/*
 * Author:
 *      Antonino Natale <antonio.natale97@hotmail.com>
 * 
 * Copyright (c) 2013-2018 Antonino Natale
 * 
 * 
 * This file is part of aPlus.
 * 
 * aPlus is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * aPlus is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with aPlus.  If not, see <http://www.gnu.org/licenses/>.
 */


#include <aplus.h>
#include <aplus/debug.h>
#include <aplus/ipc.h>

#if CONFIG_IPC

int mutex_init(mutex_t* mtx, long kind, const char* name) {
    if(unlikely(!mtx))
        return -1;

    spinlock_init(&mtx->lock);
    mtx->recursion = 0;
    mtx->kind = kind;
    mtx->owner = -1;
    mtx->name = name;
    
    return 0;
}


int mutex_lock(mutex_t* mtx) {
    if(unlikely(!mtx))
        return -1;

    if(mtx->owner != sys_getpid()) {
#if CONFIG_IPC_DEBUG
        if(likely(mtx->name))
            kprintf(LOG "[%d] %s: LOCKED from %d\n", mtx->owner, mtx->name, sys_getpid());
#endif    
        spinlock_lock(&mtx->lock);

        mtx->owner = sys_getpid();
        mtx->recursion = 0;
    } else if(mtx->kind == MTX_KIND_ERRORCHECK) {
#if CONFIG_IPC_DEBUG
        if(likely(mtx->name))
            kprintf(ERROR "[%d] %s: DEADLOCK from %d\n", mtx->owner, mtx->name, sys_getpid());
#endif
        return -1;
    }

    if(mtx->kind == MTX_KIND_RECURSIVE)
        mtx->recursion += 1;

    return 0;
}

int mutex_trylock(mutex_t* mtx) {
    if(unlikely(!mtx))
        return -1;

    if(mtx->owner != sys_getpid()) {
        if(spinlock_trylock(&mtx->lock) == E_ERR)
            return -1;

        mtx->owner = sys_getpid();
        mtx->recursion = 0;
    } else if(mtx->kind == MTX_KIND_ERRORCHECK)
        return -1;

    if(mtx->kind == MTX_KIND_RECURSIVE)
        mtx->recursion += 1;

    return 0;
}

int mutex_unlock(mutex_t* mtx) {
    if(unlikely(!mtx))
        return -1;

    if(mtx->owner == sys_getpid()) {
        if(mtx->kind == MTX_KIND_RECURSIVE)
            if(--(mtx->recursion))
                return 0;

#if CONFIG_IPC_DEBUG
        if(likely(mtx->name))
            kprintf(LOG "[%d] %s: UNLOCKED from %d\n", mtx->owner, mtx->name, sys_getpid());
#endif
    
        spinlock_unlock(&mtx->lock);
        mtx->owner = -1;
    } else if(mtx->kind == MTX_KIND_ERRORCHECK)
        return -1;

    return 0;
}

EXPORT(mutex_init);
EXPORT(mutex_lock);
EXPORT(mutex_trylock);
EXPORT(mutex_unlock);

#endif

File: kernel-rust/src/ipc/mutex.rs
/*
 * GPL3 License
 *
 * Author(s):
 *      Antonino Natale <antonio.natale97@hotmail.com>
 *
 *
 * Copyright (c) 2013-2024 Antonino Natale
 *
 * This file is part of aplus.
 *
 * aplus is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * aplus is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with aplus.  If not, see <http://www.gnu.org/licenses/>.
 */

// #[import_c_fn(arch_intr_disable)]
// #[import_c_fn(arch_intr_enable)]

use crate::bindings::{arch_intr_disable, arch_intr_enable};
use core::cell::UnsafeCell;
use core::ops::{Deref, DerefMut};
use core::sync::atomic::{AtomicBool, Ordering};

pub struct Mutex<T> {
    lock: AtomicBool,
    data: UnsafeCell<T>,
}

unsafe impl<T: Send> Sync for Mutex<T> {}

impl<T> Mutex<T> {
    pub const fn new(data: T) -> Mutex<T> {
        Mutex {
            lock: AtomicBool::new(false),
            data: UnsafeCell::new(data),
        }
    }

    pub fn lock(&self) -> MutexGuard<T> {
        while self
            .lock
            .compare_exchange(false, true, Ordering::Acquire, Ordering::Relaxed)
            .is_err()
        {
            core::hint::spin_loop();
        }
        MutexGuard {
            mutex: self,
            uninterruptible: false,
            interrupt_flags: 0,
        }
    }

    pub fn lock_uninterruptible(&self) -> MutexGuard<T> {
        let interrupt_flags = unsafe { arch_intr_disable() } as usize;
        while self
            .lock
            .compare_exchange(false, true, Ordering::Acquire, Ordering::Relaxed)
            .is_err()
        {
            core::hint::spin_loop();
        }
        MutexGuard {
            mutex: self,
            uninterruptible: true,
            interrupt_flags,
        }
    }
}

pub struct MutexGuard<'a, T> {
    mutex: &'a Mutex<T>,
    uninterruptible: bool,
    interrupt_flags: usize,
}

impl<'a, T> Drop for MutexGuard<'a, T> {
    fn drop(&mut self) {
        self.mutex.lock.store(false, Ordering::Release);
        if self.uninterruptible {
            unsafe { arch_intr_enable(self.interrupt_flags as i64) };
        }
    }
}

impl<'a, T> Deref for MutexGuard<'a, T> {
    type Target = T;

    fn deref(&self) -> &T {
        unsafe { &*self.mutex.data.get() }
    }
}

impl<'a, T> DerefMut for MutexGuard<'a, T> {
    fn deref_mut(&mut self) -> &mut T {
        unsafe { &mut *self.mutex.data.get() }
    }
}
